### ISSUE TITLE
Inconsistent parsing of relative hash references in different API v2 endpoints

### ISSUE THREAD

**Issue by dimas-b** [projectnessie/nessie#7119]
This works:
```
curl 'http://localhost:19120/api/v2/trees/main@1f7ec3eb98eef851c5e4348144ec13ce658673fabd155aefbd862be5cf07f88e~1/entries'
```
... but this:
```
curl -X 'POST' "http://localhost:19120/api/v2/trees/main@1f7ec3eb98eef851c5e4348144ec13ce658673fabd155aefbd862be5cf07f88e~1/history/commit" -d [...]
```
reports:
```

  "status" : 400,
  "reason" : "Bad Request",
  "message" : "Illegal hex character '~'",
  "errorCode" : "BAD_REQUEST",
  "serverStackTrace" : null
}
```

**Comment by adutra** [comment]
I started looking into this and fixed the error above by allowing relative specs to be passed. 

But I am wondering: does it make sense to pass a relative hash to a commit operation?

The OpenAPI description states:

> 'hash' indicates the state of contents that should be used for validating incoming changes (commits / merges / transplants)

The description also gives many examples of relative hashes:

```
          + "- -~3 - The 3rd predecessor commit from the HEAD of the default branch.\n"
          + "- name~3 - The 3rd predecessor commit from the HEAD of a branch or tag.\n"
          + "- name@hash^2 - The merge parent of the 'hash' commit of a branch or tag.\n"
          + "- -*2021-04-07T14:42:25.534748Z - The predecessor commit closest to the HEAD of the default branch for the given ISO-8601 timestamp.\n"
          + "- name*2021-04-07T14:42:25.534748Z - The predecessor commit closest to the HEAD of a branch or tag valid for the given ISO-8601 timestamp.\n"
          + "- name*1685185847230 - The predecessor commit closest to HEAD of a branch or tag valid for the given timestamp in milliseconds since epoch.\n"
```

This description doesn't look accurate. Looking at how the passed hash is used, we see that it MUST be equivalent to the current head of the branch (see [here](https://github.com/projectnessie/nessie/blob/795ba1da4cf630d40dca8b967c79c6ae2daa40eb/versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/BaseCommitHelper.java#L138)).

In short, I don't see how relative hashes could be useful here.

So here is what I propose instead:

1. Disallow relative hashes for `/history/commit`, but provide a better error message.
2. Amend the OpenAPI description and remove any mention of relative hashes for this operation.

Does that look good @dimas-b  and @snazy ?

**Comment by dimas-b** [comment]
I do not think we can require commit operations to always provide the exact HEAD of the target branch. This is not possible when multiple works commit to the same branch.

With that in mind, I do not see a big problem with allowing relative hashes for the "expected" hash during commits. It may not be very useful for automated cases, but can be handy for humans.

Also, the hash identifies a particular state of the tree in API v2 URLs. Conceptually it is at a higher level than operations on that state of the tree, so I believe we should support the same syntax for the `branch@hash~N` URL part for all sub-paths.

**Comment by dimas-b** [comment]
To clarify: I think we should allow relative hashes, then parse them, and if at that point the resultant hash is not valid for a particular operation the error should probably say that hash ABC cannot be used for "commit" (or something like that). Syntactically, I think it should be allowed.

**Comment by snazy** [comment]
I think @adutra is right here, and his proposal sounds legit.

For a commit operation, the `expectedHash` value points to the commit that the caller knows about - think: the caller knows the state of the (relevant) contents of that commit. In other words: the caller says "I want to update XYZ from the state at expectedHash to the given new value.". Then the implementation checks for conflicts _since_ that `expectedHash`.

Providing something like "HEAD minus 5 commits" does not make much sense IMHO, because HEAD is dynamic and would break the guarantees provided by `expectedHash`. In other words: it's a good source where people can shoot themselves into the foot (break consistency guarantees).

Relative hashes using timestamps might make more sense, but nothing prevents people from passing timestamps "in the future" - also considering wall-clock drift could break things in subtle ways.

**Comment by dimas-b** [comment]
I'm fine with that approach.

Still, it might be preferable to syntactically allow relative hashes even for `/history/commit` but fail later, when we determine that a relative hash was supplied. IMHO, syntax errors (like illegal character `~`) will harder to understand for end users compared to something like `using relative hashes is not supported for commits`.

**Comment by snazy** [comment]
That's a validation message issue.
